### PR TITLE
add vsts ci

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

**Comment by guolinke** [comment]
@StrikerRUS can you access vsts ci log ?

**Comment by StrikerRUS** [comment]
@guolinke 
No, 401 error:

image

File: docs/conf.py [hunk 1]
```diff
@@ -55,7 +55,7 @@
 extensions = [
     'sphinx.ext.autodoc',
     'sphinx.ext.todo',
-    'sphinx.ext.viewcode',
+#    'sphinx.ext.viewcode',
```
Comment (StrikerRUS) [inline docs/conf.py]: @guolinke So, did commenting this help to pass docs test?

File: CMakeLists.txt [hunk 2]
```diff
@@ -124,9 +124,14 @@ if(MSVC)
...
         CMAKE_CXX_FLAGS_RELWITHDEBINFO
     )
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4 /O2 /Ob2 /Oi /Ot /Oy /GL")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4 /O2 /Ob2 /Oi /Ot /Oy /GL /MP")
 else()
     SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops")
+endif()
+
+if(WIN32 AND MINGW)
```
Comment (StrikerRUS) [inline CMakeLists.txt]: Why duplicate?
Line `#108`.

File: docs/conf.py (hunk 1 above)
Comment (guolinke) [reply docs/conf.py]: yeah, it can pass now

Comment (StrikerRUS) [reply docs/conf.py]: It means that the reason was in these auto-generated links by Sphinx. To be honest, I don't know what is better: completely disable them for tests or use old LinkChecker...

image

Comment (guolinke) [reply docs/conf.py]: What will happen if disable them ?

Comment (StrikerRUS) [reply docs/conf.py]: Something like this:

image

No possibility to use links with anchors to concrete method and users cannot quickly check the code of corresponding method (`[Source]` button).

I think it's unacceptable and can be used only for tests at CI side, not at ReadTheDocs.

**Comment by StrikerRUS** [comment]
@guolinke Is there any limit of `Parallel` in VSTS? If not, then we could significantly enlarge building matrices.

And what is the decision about Travis and Appveyor after this PR will be merged? Will they be removed or left along with VSTS?

**Comment by guolinke** [comment]
it is 10 concurrent jobs now. 
I think we can keep Travis, since it is very fast.

**Comment by StrikerRUS** [comment]
>I think we can keep Travis, since it is very fast.

Can VSTS publish artifacts to GitHub Releases as Appveyor does it?

**Comment by guolinke** [comment]
10 is the max number of paid version.
It has some extensions to push back to github release, but i didn’t test that yet.

**Comment by StrikerRUS** [comment]
> 10 is the max number of paid version.

Got it! Anyway, then we can run different scenarios at Travis and VSTS and parallelize it in this way.

File: .vsts-ci/test.sh [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+if [[ $TASK == "sdist" ]]; then
+    cd ${BUILD_REPOSITORY_LOCALPATH}/python-package && python setup.py sdist || exit -1
+    pip install ${BUILD_REPOSITORY_LOCALPATH}/python-package/dist/lightgbm-$LGB_VER.tar.gz -v || exit -1
+    cp ${BUILD_REPOSITORY_LOCALPATH}/python-package/dist/lightgbm-$LGB_VER.tar.gz ${BUILD_ARTIFACTSTAGINGDIRECTORY}/lightgbm-$LGB_VER.tar.gz
+    pytest ${BUILD_REPOSITORY_LOCALPATH}/tests/python_package_test || exit -1
+    exit 0
+elif [[ $TASK == "bdist" ]]; then
+    if [[ $AGENT_OS == "Darwin" ]]; then
+        cd ${BUILD_REPOSITORY_LOCALPATH}/python-package && python setup.py bdist_wheel --plat-name=macdarwin --universal || exit -1
+        cp dist/lightgbm-$LGB_VER-py2.py3-none-macdarwin.whl ${BUILD_ARTIFACTSTAGINGDIRECTORY}/lightgbm-$LGB_VER-py2.py3-none-macosx-10.6-x86_64-macosx-10.7-x86_64-macosx-10.8-x86_64-macosx_10_9_x86_64.macosx_10_10_x86_64.macosx_10_11_x86_64.macosx_10_12_x86_64.macosx_10_13_x86_64.whl
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply .vsts-ci/test.sh]: Still different styles in delimiters:
`macosx_10.6_x86_64-macosx_10.7_x86_64` and `macosx_10_11_x86_64.macosx_10_12_x86_64`

File: docs/conf.py (hunk 1 above)
Comment (StrikerRUS) [reply docs/conf.py]: Maybe better to pin the old version of LinkChecker for now?..

**Comment by guolinke** [comment]
@StrikerRUS I think the details of vsts-ci should be accessed now, can you try it again ?

**Comment by StrikerRUS** [comment]
@guolinke Yeah, logs are visible now without logging into MS account. However, "View pipeline" (don't know what is this and should it be visible for everyone) is still giving 401 error.

File: docs/conf.py (hunk 1 above)
Comment (guolinke) [reply docs/conf.py]: okay, I don't how to pin it. can you help with that ?

Comment (StrikerRUS) [reply docs/conf.py]: Sure! Will do this.

**Comment by StrikerRUS** [comment]
Can you please grant access for restarting/canceling builds for contributors?

File: docs/conf.py (hunk 1 above)
Comment (StrikerRUS) [reply docs/conf.py]: To be honest, it's very painful to do with without possibility to cancel build and need to wait all jobs are finished (macOS is running more that 1 hour due to brew problems).

I suppose that we don't need to run `doc-test` at VSTS for now. Let Travis to perform this.

**Comment by StrikerRUS** [comment]
@guolinke Let's try it! `nekit94-12 [at] hotmail [dot] com`

I think with the buying GitHub by Microsoft the integration will be more powerful in the near future.

**Comment by StrikerRUS** [comment]
@guolinke Don't you know why there is no `Required` badge?

image

**Comment by guolinke** [comment]
invitation sent. 
I will enable the require badge after merge of this PR.

**Comment by StrikerRUS** [comment]
Hmmm.... Strange.

There is no email (in spam too), but my access to VSTS is now prohibited.

image

**Comment by guolinke** [comment]
can you access without login ? 
image
I just send a new invitation.

**Comment by StrikerRUS** [comment]
>can you access without login ?

Yeah, without logging in all thing works as before.

Seems to be the problem of delivering the mail - there is still no invitation.

**Comment by StrikerRUS** [comment]
@guolinke Homebrew problem has gone - you can rerun VSTS.

**Comment by guolinke** [comment]
it seems the vsts status cannot be updated when rebuild manually...

**Comment by StrikerRUS** [comment]
@guolinke Please invoke my invitation because now I can't see logs.

**Comment by StrikerRUS** [comment]
Also, link for details is not shown. Is it only my problem?

image

**Comment by guolinke** [comment]
@StrikerRUS 
https://docs.microsoft.com/en-us/vsts/organizations/accounts/invite-outside-users?view=vsts

It seems the external user don't have CI build access as well...

**Comment by guolinke** [comment]
@StrikerRUS can you receive invitation now ?

**Comment by StrikerRUS** [comment]
@guolinke Your invitation has arrived. Thank you very much!

I see that I can now "Queue new build", but there are no restart/stop buttons for existing builds, or I can't find them...

**Comment by guolinke** [comment]
@StrikerRUS I cannot find them either...

**Comment by guolinke** [comment]
refer to https://github.com/MicrosoftDocs/vsts-docs/issues/610, for the build from forks.

**Pull request merged**